dvc/repo/reproduce.py

```diff
@@ -101,28 +104,33 @@ def reproduce(
     active_graph = _get_active_graph(self.graph)
     active_pipelines = get_pipelines(active_graph)
 
+    stages = set()
     if pipeline or all_pipelines:
         if all_pipelines:
             pipelines = active_pipelines
         else:
-            stage = self.stage.get_target(target)
-            pipelines = [get_pipeline(active_pipelines, stage)]
+            pipelines = []
+            for target in targets:
+                stage = self.stage.get_target(target)
+                pipelines.append(get_pipeline(active_pipelines, stage))
 
-        targets = []
         for pipeline in pipelines:
             for stage in pipeline:
                 if pipeline.in_degree(stage) == 0:
-                    targets.append(stage)
+                    stages.add(stage)
     else:
-        targets = self.stage.collect(
-            target,
-            recursive=recursive,
-            graph=active_graph,
-            accept_group=accept_group,
-            glob=glob,
-        )
+        for target in targets:
+            stages.update(
+                self.stage.collect(
+                    target,
+                    recursive=recursive,
+                    graph=active_graph,
+                    accept_group=accept_group,
+                    glob=glob,
+                )
+            )
 
-    return _reproduce_stages(active_graph, targets, **kwargs)
+    return _reproduce_stages(active_graph, list(stages), **kwargs)
```

Comment:
We don't walk stages in CLI now, but we still walk the dag for each stage here. Will be addressed on top.